src/workloads/scale/MultiPlanStorm.yml

```diff
@@ -0,0 +1,93 @@
+SchemaVersion: 2018-07-01
+Owner: "@mongodb/query"
+Description: >
+  The workload tests the server under a "multi-plan storm" situation. The same query requiring a
+  multi-plan is executed by many threads, each of them triggering a multi-plan. Each of the plans in
+  the multi-plan, on each thread, is index scanning a large number of documents, while maintaining a
+  RecordId set. This causes the memory footprint to increase until the server is eventually OOM
+  killed.
+
+GlobalDefaults:
+  MaxPhases: &MaxPhases 2
+  # The amount of documents here is important, since trial-runs in the query planner "tournament"
+  # will only use ~30% of total document count increasing the number of total documents helps
+  # reproduce the issue.
+  NumDocuments: &NumDocuments 30000000
```

Comment:
@mbroadst took 15 min to get OOM killed. Should we try to make this faster?

Reply:
Yes, I think we should. That surprises me because I was able to OOM my DSI host much sooner than that (~2min?), could it be that genny is that much more efficient at generating load? Is it easy to run genny with " "explain" output to guarantee that we are hitting the multi-planning criteria we're looking for?

Reply:
Ran explain and it shows 1 winning plan + 8 rejected ones. Is this what you'd be expecting? How much RAM does your DSI host have? I will try running the JS reproducer you made to compare on Monday. 
```
{
        "explainVersion" : "2",
        "queryPlanner" : {
                "namespace" : "test.Collection0",
                "indexFilterSet" : false,
                "parsedQuery" : {
                        "$and" : [
                                {
                                        "SpaceId" : {
                                                "$eq" : 5
                                        }
                                },
                                {
                                        "Type" : {
                                                "$eq" : 6
                                        }
                                },
                                {
                                        "Tags" : {
                                                "$regex" : "erf$",
                                                "$options" : "i"
                                        }
                                },
                                {
                                        "Tags" : {
                                                "$regex" : "^per$",
                                                "$options" : "i"
                                        }
                                },
                                {
                                        "Tags" : {
                                                "$regex" : "^perf$",
                                                "$options" : "i"
                                        }
                                }
                        ]
                },
                "queryHash" : "879C963A",
                "planCacheKey" : "694AA2DC",
                "maxIndexedOrSolutionsReached" : false,
                "maxIndexedAndSolutionsReached" : false,
                "maxScansToExplodeReached" : false,
                "winningPlan" : {
                        "queryPlan" : {
                                "stage" : "FETCH",
                                "planNodeId" : 2,
                                "filter" : {
                                        "$and" : [
                                                {
                                                        "Tags" : {
                                                                "$regex" : "erf$",
                                                                "$options" : "i"
                                                        }
                                                },
                                                {
                                                        "Tags" : {
                                                                "$regex" : "^per$",
                                                                "$options" : "i"
                                                        }
                                                },
                                                {
                                                        "Tags" : {
                                                                "$regex" : "^perf$",
                                                                "$options" : "i"
                                                        }
                                                }
                                        ]
                                },
                                "inputStage" : {
                                        "stage" : "IXSCAN",
                                        "planNodeId" : 1,
                                        "keyPattern" : {
                                                "SpaceId" : 1,
                                                "Type" : 1,
                                                "Tags" : 1
                                        },
                                        "indexName" : "SpaceId_1_Type_1_Tags_1",
                                        "isMultiKey" : true,
                                        "multiKeyPaths" : {
                                                "SpaceId" : [ ],
                                                "Type" : [ ],
                                                "Tags" : [
                                                        "Tags"
                                                ]
                                        },
                                        "isUnique" : false,
                                        "isSparse" : false,
                                        "isPartial" : false,
                                        "indexVersion" : 2,
                                        "direction" : "forward",
                                        "indexBounds" : {
                                                "SpaceId" : [
                                                        "[5.0, 5.0]"
                                                ],
                                                "Type" : [
                                                        "[6.0, 6.0]"
                                                ],
                                                "Tags" : [
                                                        "[\"\", {})",
                                                        "[/erf$/i, /erf$/i]"
                                                ]
                                        }
                                }
                        },
                        "slotBasedPlan" : {
                                "slots" : "$$RESULT=s23 env: { s1 = Nothing (nothing), s3 = IndexBounds(\"field #0['SpaceId']: [5.0, 5.0], field #1['Type']: [6.0, 6.0], field #2['Tags']: [CollationKey(0x), {}), [/erf$/i, /erf$/i]\"), s7 = {\"SpaceId\" : 1, \"Type\" : 1, \"Tags\" : 1}, s11 = Nothing, s22 = true, s26 = /erf$/i, s27 = PcreRegex(/erf$/i), s28 = /^per$/i, s29 = PcreRegex(/^per$/i), s30 = /^perf$/i, s31 = PcreRegex(/^perf$/i) }",
                                "stages" : "[2] filter {(traverseF(s25, lambda(l101.0) { (((l101.0 == s26) ?: false) || (regexMatch(s27, l101.0) ?: false)) }, false) && (traverseF(s25, lambda(l102.0) { (((l102.0 == s28) ?: false) || (regexMatch(s29, l102.0) ?: false)) }, false) && traverseF(s25, lambda(l103.0) { (((l103.0 == s30) ?: false) || (regexMatch(s31, l103.0) ?: false)) }, false)))} \n[2] nlj inner [] [s17, s18, s19, s20, s21] \n    left \n        [1] unique [s17] \n        [1] branch {s22} [s17, s18, s19, s20, s21] \n        [s2, s4, s5, s6, s7] [1] ixscan_generic s3 s6 s2 s4 s5 lowPriority [] @\"c970b5fd-e18f-4430-b862-76ab5f5b8560\" @\"SpaceId_1_Type_1_Tags_1\" true \n        [s8, s14, s15, s16, s7] [1] nlj inner [] [s9, s10] \n            left \n                [1] project [s9 = getField(s12, \"l\"), s10 = getField(s12, \"h\")] \n                [1] unwind s12 s13 s11 false \n                [1] limit 1 \n                [1] coscan \n            right \n                [1] ixseek s9 s10 s16 s8 s14 s15 [] @\"c970b5fd-e18f-4430-b862-76ab5f5b8560\" @\"SpaceId_1_Type_1_Tags_1\" true \n    right \n        [2] limit 1 \n        [2] seek s17 s23 s24 s18 s19 s20 s21 none none [s25 = Tags] @\"c970b5fd-e18f-4430-b862-76ab5f5b8560\" true false \n"
                        }
                },
                "rejectedPlans" : [
                        {
                                "queryPlan" : {
                                        "stage" : "FETCH",
                                        "planNodeId" : 2,
                                        "filter" : {
                                                "$and" : [
                                                        {
                                                                "Tags" : {
                                                                        "$regex" : "erf$",
                                                                        "$options" : "i"
                                                                }
                                                        },
                                                        {
                                                                "Tags" : {
                                                                        "$regex" : "^per$",
                                                                        "$options" : "i"
                                                                }
                                                        },
                                                        {
                                                                "Tags" : {
                                                                        "$regex" : "^perf$",
                                                                        "$options" : "i"
                                                                }
                                                        },
                                                        {
                                                                "Type" : {
                                                                        "$eq" : 6
                                                                }
                                                        }
                                                ]
                                        },
                                        "inputStage" : {
                                                "stage" : "IXSCAN",
                                                "planNodeId" : 1,
                                                "keyPattern" : {
                                                        "SpaceId" : 1
                                                },
                                                "indexName" : "SpaceId_1",
                                                "isMultiKey" : false,
                                                "multiKeyPaths" : {
                                                        "SpaceId" : [ ]
                                                },
                                                "isUnique" : false,
                                                "isSparse" : false,
                                                "isPartial" : false,
                                                "indexVersion" : 2,
                                                "direction" : "forward",
                                                "indexBounds" : {
                                                        "SpaceId" : [
                                                                "[5.0, 5.0]"
                                                        ]
                                                }
                                        }
                                },
                                "slotBasedPlan" : {
                                        "slots" : "$$RESULT=s9 env: { s1 = Nothing (nothing), s2 = KS(2B0A0104), s3 = KS(2B0AFE04), s8 = {\"SpaceId\" : 1}, s13 = /erf$/i, s14 = PcreRegex(/erf$/i), s15 = /^per$/i, s16 = PcreRegex(/^per$/i), s17 = /^perf$/i, s18 = PcreRegex(/^perf$/i), s19 = 6 }",
                                        "stages" : "[2] filter {((traverseF(s11, lambda(l101.0) { (((l101.0 == s13) ?: false) || (regexMatch(s14, l101.0) ?: false)) }, false) && traverseF(s11, lambda(l102.0) { (((l102.0 == s15) ?: false) || (regexMatch(s16, l102.0) ?: false)) }, false)) && (traverseF(s11, lambda(l103.0) { (((l103.0 == s17) ?: false) || (regexMatch(s18, l103.0) ?: false)) }, false) && traverseF(s12, lambda(l104.0) { ((move(l104.0) == s19) ?: false) }, false)))} \n[2] nlj inner [] [s4, s5, s6, s7, s8] \n    left \n        [1] cfilter {(exists(s2) && exists(s3))} \n        [1] ixseek s2 s3 s7 s4 s5 s6 [] @\"c970b5fd-e18f-4430-b862-76ab5f5b8560\" @\"SpaceId_1\" true \n    right \n        [2] limit 1 \n        [2] seek s4 s9 s10 s5 s6 s7 s8 none none [s11 = Tags, s12 = Type] @\"c970b5fd-e18f-4430-b862-76ab5f5b8560\" true false \n"
                                }
                        },
                        {
                                "queryPlan" : {
                                        "stage" : "FETCH",
                                        "planNodeId" : 2,
                                        "filter" : {
                                                "$and" : [
                                                        {
                                                                "SpaceId" : {
                                                                        "$eq" : 5
                                                                }
                                                        },
                                                        {
                                                                "Tags" : {
                                                                        "$regex" : "erf$",
                                                                        "$options" : "i"
                                                                }
                                                        },
                                                        {
                                                                "Tags" : {
                                                                        "$regex" : "^per$",
                                                                        "$options" : "i"
                                                                }
                                                        },
                                                        {
                                                                "Tags" : {
                                                                        "$regex" : "^perf$",
                                                                        "$options" : "i"
                                                                }
                                                        }
                                                ]
                                        },
                                        "inputStage" : {
                                                "stage" : "IXSCAN",
                                                "planNodeId" : 1,
                                                "keyPattern" : {
                                                        "Type" : 1
                                                },
                                                "indexName" : "Type_1",
                                                "isMultiKey" : false,
                                                "multiKeyPaths" : {
                                                        "Type" : [ ]
                                                },
                                                "isUnique" : false,
                                                "isSparse" : false,
                                                "isPartial" : false,
                                                "indexVersion" : 2,
                                                "direction" : "forward",
                                                "indexBounds" : {
                                                        "Type" : [
                                                                "[6.0, 6.0]"
                                                        ]
                                                }
                                        }
                                },
                                "slotBasedPlan" : {
                                        "slots" : "$$RESULT=s9 env: { s1 = Nothing (nothing), s2 = KS(2B0C0104), s3 = KS(2B0CFE04), s8 = {\"Type\" : 1}, s13 = 5, s14 = /erf$/i, s15 = PcreRegex(/erf$/i), s16 = /^per$/i, s17 = PcreRegex(/^per$/i), s18 = /^perf$/i, s19 = PcreRegex(/^perf$/i) }",
                                        "stages" : "[2] filter {((traverseF(s11, lambda(l101.0) { ((move(l101.0) == s13) ?: false) }, false) && traverseF(s12, lambda(l102.0) { (((l102.0 == s14) ?: false) || (regexMatch(s15, l102.0) ?: false)) }, false)) && (traverseF(s12, lambda(l103.0) { (((l103.0 == s16) ?: false) || (regexMatch(s17, l103.0) ?: false)) }, false) && traverseF(s12, lambda(l104.0) { (((l104.0 == s18) ?: false) || (regexMatch(s19, l104.0) ?: false)) }, false)))} \n[2] nlj inner [] [s4, s5, s6, s7, s8] \n    left \n        [1] cfilter {(exists(s2) && exists(s3))} \n        [1] ixseek s2 s3 s7 s4 s5 s6 [] @\"c970b5fd-e18f-4430-b862-76ab5f5b8560\" @\"Type_1\" true \n    right \n        [2] limit 1 \n        [2] seek s4 s9 s10 s5 s6 s7 s8 none none [s11 = SpaceId, s12 = Tags] @\"c970b5fd-e18f-4430-b862-76ab5f5b8560\" true false \n"
                                }
                        },
                        {
                                "queryPlan" : {
                                        "stage" : "FETCH",
                                        "planNodeId" : 2,
                                        "filter" : {
                                                "$and" : [
                                                        {
                                                                "Tags" : {
                                                                        "$regex" : "erf$",
                                                                        "$options" : "i"
                                                                }
                                                        },
                                                        {
                                                                "SpaceId" : {
                                                                        "$eq" : 5
                                                                }
                                                        },
                                                        {
                                                                "Tags" : {
                                                                        "$regex" : "^per$",
                                                                        "$options" : "i"
                                                                }
                                                        },
                                                        {
                                                                "Tags" : {
                                                                        "$regex" : "^perf$",
                                                                        "$options" : "i"
                                                                }
                                                        },
                                                        {
                                                                "Type" : {
                                                                        "$eq" : 6
                                                                }
                                                        }
                                                ]
                                        },
                                        "inputStage" : {
                                                "stage" : "IXSCAN",
                                                "planNodeId" : 1,
                                                "keyPattern" : {
                                                        "Tags" : 1
                                                },
                                                "indexName" : "Tags_1",
                                                "isMultiKey" : true,
                                                "multiKeyPaths" : {
                                                        "Tags" : [
                                                                "Tags"
                                                        ]
                                                },
                                                "isUnique" : false,
                                                "isSparse" : false,
                                                "isPartial" : false,
                                                "indexVersion" : 2,
                                                "direction" : "forward",
                                                "indexBounds" : {
                                                        "Tags" : [
                                                                "[\"\", {})",
                                                                "[/erf$/i, /erf$/i]"
                                                        ]
                                                }
                                        }
                                },
                                "slotBasedPlan" : {
                                        "slots" : "$$RESULT=s23 env: { s1 = Nothing (nothing), s3 = IndexBounds(\"field #0['Tags']: [CollationKey(0x), {}), [/erf$/i, /erf$/i]\"), s7 = {\"Tags\" : 1}, s11 = Nothing, s22 = true, s28 = /erf$/i, s29 = PcreRegex(/erf$/i), s30 = 5, s31 = /^per$/i, s32 = PcreRegex(/^per$/i), s33 = /^perf$/i, s34 = PcreRegex(/^perf$/i), s35 = 6 }",
                                        "stages" : "[2] filter {((traverseF(s26, lambda(l101.0) { (((l101.0 == s28) ?: false) || (regexMatch(s29, l101.0) ?: false)) }, false) && traverseF(s25, lambda(l102.0) { ((move(l102.0) == s30) ?: false) }, false)) && (traverseF(s26, lambda(l103.0) { (((l103.0 == s31) ?: false) || (regexMatch(s32, l103.0) ?: false)) }, false) && (traverseF(s26, lambda(l104.0) { (((l104.0 == s33) ?: false) || (regexMatch(s34, l104.0) ?: false)) }, false) && traverseF(s27, lambda(l105.0) { ((move(l105.0) == s35) ?: false) }, false))))} \n[2] nlj inner [] [s17, s18, s19, s20, s21] \n    left \n        [1] unique [s17] \n        [1] branch {s22} [s17, s18, s19, s20, s21] \n        [s2, s4, s5, s6, s7] [1] ixscan_generic s3 s6 s2 s4 s5 lowPriority [] @\"c970b5fd-e18f-4430-b862-76ab5f5b8560\" @\"Tags_1\" true \n        [s8, s14, s15, s16, s7] [1] nlj inner [] [s9, s10] \n            left \n                [1] project [s9 = getField(s12, \"l\"), s10 = getField(s12, \"h\")] \n                [1] unwind s12 s13 s11 false \n                [1] limit 1 \n                [1] coscan \n            right \n                [1] ixseek s9 s10 s16 s8 s14 s15 [] @\"c970b5fd-e18f-4430-b862-76ab5f5b8560\" @\"Tags_1\" true \n    right \n        [2] limit 1 \n        [2] seek s17 s23 s24 s18 s19 s20 s21 none none [s25 = SpaceId, s26 = Tags, s27 = Type] @\"c970b5fd-e18f-4430-b862-76ab5f5b8560\" true false \n"
                                }
                        },
                        {
                                "queryPlan" : {
                                        "stage" : "FETCH",
                                        "planNodeId" : 2,
                                        "filter" : {
                                                "$and" : [
                                                        {
                                                                "Tags" : {
                                                                        "$regex" : "^per$",
                                                                        "$options" : "i"
                                                                }
                                                        },
                                                        {
                                                                "SpaceId" : {
                                                                        "$eq" : 5
                                                                }
                                                        },
                                                        {
                                                                "Tags" : {
                                                                        "$regex" : "erf$",
                                                                        "$options" : "i"
                                                                }
                                                        },
                                                        {
                                                                "Tags" : {
                                                                        "$regex" : "^perf$",
                                                                        "$options" : "i"
                                                                }
                                                        },
                                                        {
                                                                "Type" : {
                                                                        "$eq" : 6
                                                                }
                                                        }
                                                ]
                                        },
                                        "inputStage" : {
                                                "stage" : "IXSCAN",
                                                "planNodeId" : 1,
                                                "keyPattern" : {
                                                        "Tags" : 1
                                                },
                                                "indexName" : "Tags_1",
                                                "isMultiKey" : true,
                                                "multiKeyPaths" : {
                                                        "Tags" : [
                                                                "Tags"
                                                        ]
                                                },
                                                "isUnique" : false,
                                                "isSparse" : false,
                                                "isPartial" : false,
                                                "indexVersion" : 2,
                                                "direction" : "forward",
                                                "indexBounds" : {
                                                        "Tags" : [
                                                                "[\"\", {})",
                                                                "[/^per$/i, /^per$/i]"
                                                        ]
                                                }
                                        }
                                },
                                "slotBasedPlan" : {
                                        "slots" : "$$RESULT=s23 env: { s1 = Nothing (nothing), s3 = IndexBounds(\"field #0['Tags']: [CollationKey(0x), {}), [/^per$/i, /^per$/i]\"), s7 = {\"Tags\" : 1}, s11 = Nothing, s22 = true, s28 = /^per$/i, s29 = PcreRegex(/^per$/i), s30 = 5, s31 = /erf$/i, s32 = PcreRegex(/erf$/i), s33 = /^perf$/i, s34 = PcreRegex(/^perf$/i), s35 = 6 }",
                                        "stages" : "[2] filter {((traverseF(s26, lambda(l101.0) { (((l101.0 == s28) ?: false) || (regexMatch(s29, l101.0) ?: false)) }, false) && traverseF(s25, lambda(l102.0) { ((move(l102.0) == s30) ?: false) }, false)) && (traverseF(s26, lambda(l103.0) { (((l103.0 == s31) ?: false) || (regexMatch(s32, l103.0) ?: false)) }, false) && (traverseF(s26, lambda(l104.0) { (((l104.0 == s33) ?: false) || (regexMatch(s34, l104.0) ?: false)) }, false) && traverseF(s27, lambda(l105.0) { ((move(l105.0) == s35) ?: false) }, false))))} \n[2] nlj inner [] [s17, s18, s19, s20, s21] \n    left \n        [1] unique [s17] \n        [1] branch {s22} [s17, s18, s19, s20, s21] \n        [s2, s4, s5, s6, s7] [1] ixscan_generic s3 s6 s2 s4 s5 lowPriority [] @\"c970b5fd-e18f-4430-b862-76ab5f5b8560\" @\"Tags_1\" true \n        [s8, s14, s15, s16, s7] [1] nlj inner [] [s9, s10] \n            left \n                [1] project [s9 = getField(s12, \"l\"), s10 = getField(s12, \"h\")] \n                [1] unwind s12 s13 s11 false \n                [1] limit 1 \n                [1] coscan \n            right \n                [1] ixseek s9 s10 s16 s8 s14 s15 [] @\"c970b5fd-e18f-4430-b862-76ab5f5b8560\" @\"Tags_1\" true \n    right \n        [2] limit 1 \n        [2] seek s17 s23 s24 s18 s19 s20 s21 none none [s25 = SpaceId, s26 = Tags, s27 = Type] @\"c970b5fd-e18f-4430-b862-76ab5f5b8560\" true false \n"
                                }
                        },
                        {
                                "queryPlan" : {
                                        "stage" : "FETCH",
                                        "planNodeId" : 2,
                                        "filter" : {
                                                "$and" : [
                                                        {
                                                                "Tags" : {
                                                                        "$regex" : "^perf$",
                                                                        "$options" : "i"
                                                                }
                                                        },
                                                        {
                                                                "SpaceId" : {
                                                                        "$eq" : 5
                                                                }
                                                        },
                                                        {
                                                                "Tags" : {
                                                                        "$regex" : "erf$",
                                                                        "$options" : "i"
                                                                }
                                                        },
                                                        {
                                                                "Tags" : {
                                                                        "$regex" : "^per$",
                                                                        "$options" : "i"
                                                                }
                                                        },
                                                        {
                                                                "Type" : {
                                                                        "$eq" : 6
                                                                }
                                                        }
                                                ]
                                        },
                                        "inputStage" : {
                                                "stage" : "IXSCAN",
                                                "planNodeId" : 1,
                                                "keyPattern" : {
                                                        "Tags" : 1
                                                },
                                                "indexName" : "Tags_1",
                                                "isMultiKey" : true,
                                                "multiKeyPaths" : {
                                                        "Tags" : [
                                                                "Tags"
                                                        ]
                                                },
                                                "isUnique" : false,
                                                "isSparse" : false,
                                                "isPartial" : false,
                                                "indexVersion" : 2,
                                                "direction" : "forward",
                                                "indexBounds" : {
                                                        "Tags" : [
                                                                "[\"\", {})",
                                                                "[/^perf$/i, /^perf$/i]"
                                                        ]
                                                }
                                        }
                                },
                                "slotBasedPlan" : {
                                        "slots" : "$$RESULT=s23 env: { s1 = Nothing (nothing), s3 = IndexBounds(\"field #0['Tags']: [CollationKey(0x), {}), [/^perf$/i, /^perf$/i]\"), s7 = {\"Tags\" : 1}, s11 = Nothing, s22 = true, s28 = /^perf$/i, s29 = PcreRegex(/^perf$/i), s30 = 5, s31 = /erf$/i, s32 = PcreRegex(/erf$/i), s33 = /^per$/i, s34 = PcreRegex(/^per$/i), s35 = 6 }",
                                        "stages" : "[2] filter {((traverseF(s26, lambda(l101.0) { (((l101.0 == s28) ?: false) || (regexMatch(s29, l101.0) ?: false)) }, false) && traverseF(s25, lambda(l102.0) { ((move(l102.0) == s30) ?: false) }, false)) && (traverseF(s26, lambda(l103.0) { (((l103.0 == s31) ?: false) || (regexMatch(s32, l103.0) ?: false)) }, false) && (traverseF(s26, lambda(l104.0) { (((l104.0 == s33) ?: false) || (regexMatch(s34, l104.0) ?: false)) }, false) && traverseF(s27, lambda(l105.0) { ((move(l105.0) == s35) ?: false) }, false))))} \n[2] nlj inner [] [s17, s18, s19, s20, s21] \n    left \n        [1] unique [s17] \n        [1] branch {s22} [s17, s18, s19, s20, s21] \n        [s2, s4, s5, s6, s7] [1] ixscan_generic s3 s6 s2 s4 s5 lowPriority [] @\"c970b5fd-e18f-4430-b862-76ab5f5b8560\" @\"Tags_1\" true \n        [s8, s14, s15, s16, s7] [1] nlj inner [] [s9, s10] \n            left \n                [1] project [s9 = getField(s12, \"l\"), s10 = getField(s12, \"h\")] \n                [1] unwind s12 s13 s11 false \n                [1] limit 1 \n                [1] coscan \n            right \n                [1] ixseek s9 s10 s16 s8 s14 s15 [] @\"c970b5fd-e18f-4430-b862-76ab5f5b8560\" @\"Tags_1\" true \n    right \n        [2] limit 1 \n        [2] seek s17 s23 s24 s18 s19 s20 s21 none none [s25 = SpaceId, s26 = Tags, s27 = Type] @\"c970b5fd-e18f-4430-b862-76ab5f5b8560\" true false \n"
                                }
                        },
                        {
                                "queryPlan" : {
                                        "stage" : "FETCH",
                                        "planNodeId" : 2,
                                        "filter" : {
                                                "$and" : [
                                                        {
                                                                "Tags" : {
                                                                        "$regex" : "erf$",
                                                                        "$options" : "i"
                                                                }
                                                        },
                                                        {
                                                                "Tags" : {
                                                                        "$regex" : "^per$",
                                                                        "$options" : "i"
                                                                }
                                                        },
                                                        {
                                                                "Tags" : {
                                                                        "$regex" : "^perf$",
                                                                        "$options" : "i"
                                                                }
                                                        }
                                                ]
                                        },
                                        "inputStage" : {
                                                "stage" : "IXSCAN",
                                                "planNodeId" : 1,
                                                "keyPattern" : {
                                                        "SpaceId" : 1,
                                                        "Type" : 1
                                                },
                                                "indexName" : "SpaceId_1_Type_1",
                                                "isMultiKey" : false,
                                                "multiKeyPaths" : {
                                                        "SpaceId" : [ ],
                                                        "Type" : [ ]
                                                },
                                                "isUnique" : false,
                                                "isSparse" : false,
                                                "isPartial" : false,
                                                "indexVersion" : 2,
                                                "direction" : "forward",
                                                "indexBounds" : {
                                                        "SpaceId" : [
                                                                "[5.0, 5.0]"
                                                        ],
                                                        "Type" : [
                                                                "[6.0, 6.0]"
                                                        ]
                                                }
                                        }
                                },
                                "slotBasedPlan" : {
                                        "slots" : "$$RESULT=s9 env: { s1 = Nothing (nothing), s2 = KS(2B0A2B0C0104), s3 = KS(2B0A2B0CFE04), s8 = {\"SpaceId\" : 1, \"Type\" : 1}, s12 = /erf$/i, s13 = PcreRegex(/erf$/i), s14 = /^per$/i, s15 = PcreRegex(/^per$/i), s16 = /^perf$/i, s17 = PcreRegex(/^perf$/i) }",
                                        "stages" : "[2] filter {(traverseF(s11, lambda(l101.0) { (((l101.0 == s12) ?: false) || (regexMatch(s13, l101.0) ?: false)) }, false) && (traverseF(s11, lambda(l102.0) { (((l102.0 == s14) ?: false) || (regexMatch(s15, l102.0) ?: false)) }, false) && traverseF(s11, lambda(l103.0) { (((l103.0 == s16) ?: false) || (regexMatch(s17, l103.0) ?: false)) }, false)))} \n[2] nlj inner [] [s4, s5, s6, s7, s8] \n    left \n        [1] cfilter {(exists(s2) && exists(s3))} \n        [1] ixseek s2 s3 s7 s4 s5 s6 [] @\"c970b5fd-e18f-4430-b862-76ab5f5b8560\" @\"SpaceId_1_Type_1\" true \n    right \n        [2] limit 1 \n        [2] seek s4 s9 s10 s5 s6 s7 s8 none none [s11 = Tags] @\"c970b5fd-e18f-4430-b862-76ab5f5b8560\" true false \n"
                                }
                        },
                        {
                                "queryPlan" : {
                                        "stage" : "FETCH",
                                        "planNodeId" : 2,
                                        "filter" : {
                                                "$and" : [
                                                        {
                                                                "Tags" : {
                                                                        "$regex" : "erf$",
                                                                        "$options" : "i"
                                                                }
                                                        },
                                                        {
                                                                "Tags" : {
                                                                        "$regex" : "^per$",
                                                                        "$options" : "i"
                                                                }
                                                        },
                                                        {
                                                                "Tags" : {
                                                                        "$regex" : "^perf$",
                                                                        "$options" : "i"
                                                                }
                                                        },
                                                        {
                                                                "Type" : {
                                                                        "$eq" : 6
                                                                }
                                                        }
                                                ]
                                        },
                                        "inputStage" : {
                                                "stage" : "IXSCAN",
                                                "planNodeId" : 1,
                                                "keyPattern" : {
                                                        "SpaceId" : 1,
                                                        "Tags" : 1
                                                },
                                                "indexName" : "SpaceId_1_Tags_1",
                                                "isMultiKey" : true,
                                                "multiKeyPaths" : {
                                                        "SpaceId" : [ ],
                                                        "Tags" : [
                                                                "Tags"
                                                        ]
                                                },
                                                "isUnique" : false,
                                                "isSparse" : false,
                                                "isPartial" : false,
                                                "indexVersion" : 2,
                                                "direction" : "forward",
                                                "indexBounds" : {
                                                        "SpaceId" : [
                                                                "[5.0, 5.0]"
                                                        ],
                                                        "Tags" : [
                                                                "[\"\", {})",
                                                                "[/erf$/i, /erf$/i]"
                                                        ]
                                                }
                                        }
                                },
                                "slotBasedPlan" : {
                                        "slots" : "$$RESULT=s23 env: { s1 = Nothing (nothing), s3 = IndexBounds(\"field #0['SpaceId']: [5.0, 5.0], field #1['Tags']: [CollationKey(0x), {}), [/erf$/i, /erf$/i]\"), s7 = {\"SpaceId\" : 1, \"Tags\" : 1}, s11 = Nothing, s22 = true, s27 = /erf$/i, s28 = PcreRegex(/erf$/i), s29 = /^per$/i, s30 = PcreRegex(/^per$/i), s31 = /^perf$/i, s32 = PcreRegex(/^perf$/i), s33 = 6 }",
                                        "stages" : "[2] filter {((traverseF(s25, lambda(l101.0) { (((l101.0 == s27) ?: false) || (regexMatch(s28, l101.0) ?: false)) }, false) && traverseF(s25, lambda(l102.0) { (((l102.0 == s29) ?: false) || (regexMatch(s30, l102.0) ?: false)) }, false)) && (traverseF(s25, lambda(l103.0) { (((l103.0 == s31) ?: false) || (regexMatch(s32, l103.0) ?: false)) }, false) && traverseF(s26, lambda(l104.0) { ((move(l104.0) == s33) ?: false) }, false)))} \n[2] nlj inner [] [s17, s18, s19, s20, s21] \n    left \n        [1] unique [s17] \n        [1] branch {s22} [s17, s18, s19, s20, s21] \n        [s2, s4, s5, s6, s7] [1] ixscan_generic s3 s6 s2 s4 s5 lowPriority [] @\"c970b5fd-e18f-4430-b862-76ab5f5b8560\" @\"SpaceId_1_Tags_1\" true \n        [s8, s14, s15, s16, s7] [1] nlj inner [] [s9, s10] \n            left \n                [1] project [s9 = getField(s12, \"l\"), s10 = getField(s12, \"h\")] \n                [1] unwind s12 s13 s11 false \n                [1] limit 1 \n                [1] coscan \n            right \n                [1] ixseek s9 s10 s16 s8 s14 s15 [] @\"c970b5fd-e18f-4430-b862-76ab5f5b8560\" @\"SpaceId_1_Tags_1\" true \n    right \n        [2] limit 1 \n        [2] seek s17 s23 s24 s18 s19 s20 s21 none none [s25 = Tags, s26 = Type] @\"c970b5fd-e18f-4430-b862-76ab5f5b8560\" true false \n"
                                }
                        },
                        {
                                "queryPlan" : {
                                        "stage" : "FETCH",
                                        "planNodeId" : 2,
                                        "filter" : {
                                                "$and" : [
                                                        {
                                                                "Tags" : {
                                                                        "$regex" : "erf$",
                                                                        "$options" : "i"
                                                                }
                                                        },
                                                        {
                                                                "SpaceId" : {
                                                                        "$eq" : 5
                                                                }
                                                        },
                                                        {
                                                                "Tags" : {
                                                                        "$regex" : "^per$",
                                                                        "$options" : "i"
                                                                }
                                                        },
                                                        {
                                                                "Tags" : {
                                                                        "$regex" : "^perf$",
                                                                        "$options" : "i"
                                                                }
                                                        }
                                                ]
                                        },
                                        "inputStage" : {
                                                "stage" : "IXSCAN",
                                                "planNodeId" : 1,
                                                "keyPattern" : {
                                                        "Type" : 1,
                                                        "Tags" : 1
                                                },
                                                "indexName" : "Type_1_Tags_1",
                                                "isMultiKey" : true,
                                                "multiKeyPaths" : {
                                                        "Type" : [ ],
                                                        "Tags" : [
                                                                "Tags"
                                                        ]
                                                },
                                                "isUnique" : false,
                                                "isSparse" : false,
                                                "isPartial" : false,
                                                "indexVersion" : 2,
                                                "direction" : "forward",
                                                "indexBounds" : {
                                                        "Type" : [
                                                                "[6.0, 6.0]"
                                                        ],
                                                        "Tags" : [
                                                                "[\"\", {})",
                                                                "[/erf$/i, /erf$/i]"
                                                        ]
                                                }
                                        }
                                },
                                "slotBasedPlan" : {
                                        "slots" : "$$RESULT=s23 env: { s1 = Nothing (nothing), s3 = IndexBounds(\"field #0['Type']: [6.0, 6.0], field #1['Tags']: [CollationKey(0x), {}), [/erf$/i, /erf$/i]\"), s7 = {\"Type\" : 1, \"Tags\" : 1}, s11 = Nothing, s22 = true, s27 = /erf$/i, s28 = PcreRegex(/erf$/i), s29 = 5, s30 = /^per$/i, s31 = PcreRegex(/^per$/i), s32 = /^perf$/i, s33 = PcreRegex(/^perf$/i) }",
                                        "stages" : "[2] filter {((traverseF(s26, lambda(l101.0) { (((l101.0 == s27) ?: false) || (regexMatch(s28, l101.0) ?: false)) }, false) && traverseF(s25, lambda(l102.0) { ((move(l102.0) == s29) ?: false) }, false)) && (traverseF(s26, lambda(l103.0) { (((l103.0 == s30) ?: false) || (regexMatch(s31, l103.0) ?: false)) }, false) && traverseF(s26, lambda(l104.0) { (((l104.0 == s32) ?: false) || (regexMatch(s33, l104.0) ?: false)) }, false)))} \n[2] nlj inner [] [s17, s18, s19, s20, s21] \n    left \n        [1] unique [s17] \n        [1] branch {s22} [s17, s18, s19, s20, s21] \n        [s2, s4, s5, s6, s7] [1] ixscan_generic s3 s6 s2 s4 s5 lowPriority [] @\"c970b5fd-e18f-4430-b862-76ab5f5b8560\" @\"Type_1_Tags_1\" true \n        [s8, s14, s15, s16, s7] [1] nlj inner [] [s9, s10] \n            left \n                [1] project [s9 = getField(s12, \"l\"), s10 = getField(s12, \"h\")] \n                [1] unwind s12 s13 s11 false \n                [1] limit 1 \n                [1] coscan \n            right \n                [1] ixseek s9 s10 s16 s8 s14 s15 [] @\"c970b5fd-e18f-4430-b862-76ab5f5b8560\" @\"Type_1_Tags_1\" true \n    right \n        [2] limit 1 \n        [2] seek s17 s23 s24 s18 s19 s20 s21 none none [s25 = SpaceId, s26 = Tags] @\"c970b5fd-e18f-4430-b862-76ab5f5b8560\" true false \n"
                                }
                        }
                ]
        },
        "command" : {
                "find" : "Collection0",
                "filter" : {
                        "$and" : [
                                {
                                        "SpaceId" : {
                                                "$eq" : 5
                                        }
                                },
                                {
                                        "Type" : {
                                                "$eq" : 6
                                        }
                                },
                                {
                                        "Tags" : {
                                                "$regex" : /erf$/,
                                                "$options" : "i"
                                        }
                                },
                                {
                                        "Tags" : {
                                                "$regex" : /^per$/,
                                                "$options" : "i"
                                        }
                                },
                                {
                                        "Tags" : {
                                                "$regex" : /^perf$/,
                                                "$options" : "i"
                                        }
                                }
                        ]
                },
                "$db" : "test"
        },
        "serverInfo" : {
                "host" : "ip-10-2-0-200.ec2.internal",
                "port" : 27017,
                "version" : "7.3.0-alpha-820-g26e5fb7-sys-perf-patch-65cf5820d1fe0731910b031f",
                "gitVersion" : "26e5fb7e890b6477fa823934e799841da0eecc40"
        },
        "serverParameters" : {
                "internalQueryFacetBufferSizeBytes" : 104857600,
                "internalQueryFacetMaxOutputDocSizeBytes" : 104857600,
                "internalLookupStageIntermediateDocumentMaxSizeBytes" : 104857600,
                "internalDocumentSourceGroupMaxMemoryBytes" : 104857600,
                "internalQueryMaxBlockingSortMemoryUsageBytes" : 104857600,
                "internalQueryProhibitBlockingMergeOnMongoS" : 0,
                "internalQueryMaxAddToSetBytes" : 104857600,
                "internalDocumentSourceSetWindowFieldsMaxMemoryBytes" : 104857600,
                "internalQueryFrameworkControl" : "trySbeEngine"
        },
        "ok" : 1,
        "$clusterTime" : {
                "clusterTime" : Timestamp(1708106631, 1),
                "signature" : {
                        "hash" : BinData(0,"AAAAAAAAAAAAAAAAAAAAAAAAAAA="),
                        "keyId" : NumberLong(0)
                }
        },
        "operationTime" : Timestamp(1708106631, 1)
}
```

Reply:
it looks like about the right number of rejected plans, can you try running with `explain('allPlansExecution')` which will return specific execution statistics? We're specifically looking for very high numbers of `dupesTested` on the rejected plans

Reply:
I'm not remembering that my repro was against 4.4, so it's possible that we've just become more efficient at this particular query in recent years.